initrd/bin/config-gui.sh

```diff
@@ -141,18 +172,305 @@ while true; do
+          set_config /etc/config.user "CONFIG_ROOT_CHECK_AT_BOOT" "y"
+          combine_configs
+
+          # check that root hash file exists
+          if [ ! -f ${ROOT_HASH_FILE} ]; then
+            if (whiptail --title 'Generate Root Hash File' \
+                --yesno "\nNo root hash file exists.
+                        \nWould you like to create the initial hash file now?" 0 80) then
+                root-hashes-gui.sh -n
+              fi
+          fi
+
+          whiptail --title 'Config change successful' \
+            --msgbox "The root device will be checked at each boot." 0 80
+
+        fi
+      else
+        if (whiptail --title 'Disable Root Hash Check at Boot?' \
+             --yesno "This will disable checking root hashes each time you boot.
+                    \n\nDo you want to proceed?" 0 80) then
+
+          set_config /etc/config.user "CONFIG_ROOT_CHECK_AT_BOOT" "n"
+          combine_configs
+
+          whiptail --title 'Config change successful' \
+            --msgbox "The root device will not be checked at each boot." 0 80
+        fi
+      fi
+    ;;
+    "P" )
+      if [ "$CONFIG_RESTRICTED_BOOT" = "y" ]; then
+          whiptail $BG_COLOR_ERROR --title 'Restricted Boot Active' \
+            --msgbox "Disable Restricted Boot to enable Basic Mode." 0 80
+      elif [ "$CONFIG_BASIC" != "y" ]; then
+        if (whiptail --title "Enable $CONFIG_BRAND_NAME Basic Mode?" \
+             --yesno "This will remove all signature checking on the firmware
+                    \nand boot files, and disable use of the Librem Key.
+                    \n\nDo you want to proceed?" 0 80) then
+
+          set_user_config "CONFIG_BASIC" "y"
+
+          whiptail --title 'Config change successful' \
+            --msgbox "$CONFIG_BRAND_NAME Basic mode enabled;\nsave the config change and reboot for it to go into effect." 0 80
+
+        fi
+      else
+        if (whiptail --title "Disable $CONFIG_BRAND_NAME Basic Mode?" \
+             --yesno "This will enable all signature checking on the firmware
+                    \nand boot files, and enable use of the Librem Key.
+                    \n\nDo you want to proceed?" 0 80) then
+
+          set_user_config "CONFIG_BASIC" "n"
+
+          whiptail --title 'Config change successful' \
+            --msgbox "$CONFIG_BRAND_NAME Basic mode has been disabled;\nsave the config change and reboot for it to go into effect." 0 80
+        fi
+      fi
+    ;;
+    "L" )
+      if [ "$CONFIG_RESTRICTED_BOOT" != "y" ]; then
+        if (whiptail --title 'Enable Restricted Boot Mode?' \
+             --yesno "This will disable booting from any unsigned files,
+                    \nincluding kernels that have not yet been signed,
+                    \n.isos without signatures, raw USB disks,
+                    \nand will disable failsafe boot mode.
+                    \n\nThis will also disable the recovery console.
+                    \n\nDo you want to proceed?" 0 80) then
```

Comment:
@JonathonHall-Purism an important feature of Restricted boot is not advertised here: with restricted mode enabled, a user won't be able to update firmware unless restricted boot is disabled, as can be seen from flash-gui.sh when an upgrade is attempted. This warning should also be given in the config gui prompt for enablement.

I'm impressed with the effectiveness of restricted boot here and i'm questioning myself on next steps (issues to be opened for implementation):
- Should restricted boot be configured as default under Heads?
- Should theUSB boot option for ISO not having a detached signature be prompted to be detached signed and verified with user's keyring?

Reply:
>This warning should also be given in the config gui prompt for enablement.

Agree :+1:

This is already a somewhat long/wordy prompt, but I think we can capture all of this still.

For comparison, existing wording:
```
This will disable booting from any unsigned files,
including kernels that have not yet been signed,
isos without signatures, raw USB disks,
and will disable failsafe boot mode.

This will also disable the recovery console.

Do you want to proceed?
```

How about:
```
Restricted Boot allows booting:
* Signed installed OS
* Signed ISOs from USB

All other boot methods are blocked.  Recovery console and firmware updates will
be blocked.

Restricted boot can be disabled at any time.  This resets TOTP/HOTP so it is
evident that Restricted Boot was disabled.

Do you want to proceed?
```

It's obviously not short but I think it's more accessible:
* most important information first
* avoids long/complex sentences
* adds that you can still turn it off in case you are unsure this is what you want

I waiver about including the "resets TOTP/HOTP" bit, that's perhaps a bit more information than is really needed, but OTOH there is a really obvious question about "what good is it if anyone can disable it" and that's the answer, disabling it creates evidence.

Reply:
>Should restricted boot be configured as default under Heads?

In principle this is easy to do just by setting `export CONFIG_RESTRICTED_BOOT=y` in a board config if you want a board to have it enabled by default.  Should still work to configure it from the GUI since that overrides board config.

I also offer some "preconfigured" ROMs (where I've just injected a config.user ahead of time) for specific configurations like basic, Mini basic + auto power-on + USB autoboot (headless config), blob jail, etc.  This works really well IMO, coreboot_util.sh offers a choice of configs if you are switching to PB from SeaBIOS.  I don't offer them on PB upgrades because by default it preserves your existing settings, but you actually can flash the preconfigures with the "erasing settings" method and it works.

>Should the USB boot option for ISO not having a detached signature be prompted to be detached signed and verified with user's keyring?

I do think there's a use case here.  IMO, we should think about workflows so we are not bombarding with too many serial questions of "do you want to X", "looks like you are trying to Y".

I think it'd be a good solution to have a separate "signing GUI" to sign an ISO with your key, which would fulfill this.  It'd also be great to be able to configure which OS keys are trusted with RB; maybe you would prefer to only allow booting ISOs you yourself have signed.

Reply:
I'm an advocate of bullet points detailing what is gonna be lost and active. :/
I like your suggestion, I say let's go with it for simplicity I guess. As you said numerous times, if its too long people don't read. It's sad, but this is what it is.

Reply:
>I think it'd be a good solution to have a separate "signing GUI" to sign an ISO with your key, which would fulfill this. It'd also be great to be able to configure which OS keys are trusted with RB; maybe you would prefer to only allow booting ISOs you yourself have signed.

Yes. This implies that users will have verified hashes externally, or have the GUI to offer the possibility to generate sha256sum hashes prior of detach signing in case hashes were not already verified on thumb drive. I won't insist enough here, but outside of hashes, so many users are using cheap dongles, and people tend to verify hashes on computer, not on thumb drive. So signing an unverified iso hash is kind of dumb.

Please open issue for tracking on this one. I think it will be beneficial to all and easy to do.

Reply:
I think this is pretty reasonable.  I took out the double spacing as it made the message monstrously tall in fbwhiptail, and the output is already sort of double-spaced.

Reworded the "disable" prompt similarly, let me know if you have any comments on that.

![Screenshot_20230711_155809](https://github.com/osresearch/heads/assets/109107914/352d0589-e4d7-4b01-a3b5-5f13632e80a9)

![Screenshot_20230711_155822](https://github.com/osresearch/heads/assets/109107914/e954ba71-ca29-4ff6-a771-13ee060b1f29)